src/python/pants/engine/legacy/graph.py

```diff
@@ -360,24 +361,39 @@ def transitive_dependents_of_addresses(self, addresses):
         return result
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass
 class HydratedTarget:
     """A wrapper for a fully hydrated TargetAdaptor object.
 
+    Why have this type if all information is stored in the underlying TargetAdaptor? We need it for
+    the type-driven graph to work properly. A HydratedTarget is a generic wrapper around different
+    target classes, whereas a TargetAdaptor has different types like PythonTestsAdaptor and
+    PythonLibraryAdaptor. We need to be able to work with targets both generically and depending
+    on their target type (i.e. unions).
+
     Transitive graph walks collect ordered sets of TransitiveHydratedTargets which involve a huge
-    amount of hashing: we implement eq/hash via direct usage of an Address field to speed that up.
+    amount of hashing: we implement eq/hash via a private Address field to speed that up.
     """
 
-    address: Address
     adaptor: TargetAdaptor
-    dependencies: Tuple[Address, ...]
 
-    @property
-    def addresses(self) -> Tuple:
-        return self.dependencies
+    def __init__(self, adaptor: TargetAdaptor) -> None:
+        self.adaptor = adaptor
+        # This field is set for efficient lookup of the address in `__hash__` and `__eq__` during
+        # graph walks. Directly accessing this field, rather than using `adaptor.address` is about
+        # 10x faster. However, rules should still use `adaptor.address` for clarity and because
+        # they do not need to access the address as frequently, so it would be an over-optimization
+        # (we're talking about 0.000_000_1 vs. 0.000_001 seconds here).
+        self._address = adaptor.address
```

Comment:
NB: as mentioned here, an alternative would be to always surface `Address` as a public field. We could have this custom `__init__` so that call sites get the benefits of only ever passing a `TargetAdaptor` but still have the same fields as before (smaller diff / less verbosity).

Why do I think we should keep this field private? Reinforce the conceptual distinction between `HydratedTarget` vs. `TargetAdaptor`. `TargetAdaptor` is what actually stores all the information. `HydratedTarget` is simply a wrapper so that we have a generic reference for the engine. This is similar to how `AddressWithOrigin` is simply an `Address` augmented with the origin information.